### DISCUSSION TITLE
Plans for the library

### DISCUSSION THREAD

**Discussion by technosophos** [GravityPDF/querypath#35] [General]
What are the plans for maintenance and improvement of the library?

I haven't worked on QueryPath for a long, long time now. And if you are interested in doing the long-term maintenance on this fork, then I might be able to help point users this way.

Thanks for breathing some new life into a project I've always been proud of, but which has diverged from my longer-term career path!

**Comment by jakejackson1** [comment]
Hey @technosophos, thanks for dropping by! It's a real pleasure to have the lead developer stop in and say hello. 👋

The primary goal for this fork is to ensure the library continues functioning as PHP evolves, and  that any security issues found/reported are promptly fixed. On the one hand, this is somewhat self-serving, [as my company makes use of QueryPath in our primary product](https://github.com/gravityPDF/gravity-pdf/). But on the other, that means we have a vested interest in ensuring QueryPath lives on. 

The long term and secondary goals for the fork is to:

1. Refactor and modernize the codebase, while maintaining backwards compatibility as much as possible
2. Write documentation to covers common use cases and pitfalls, so new developers can get started fast
4. Facilitate and foster collaboration with the QueryPath community, so that the library doesn't live and die with one person or organization. 

As for why choose "us" and this fork: judge for yourself, but I feel my company and I have a good track record when it comes to contributing to open source software. We aren't a large company by any means (10), but we are a stable, profitable business who has been around for over 10 years now.

Talk is cheap though, so here's how we've already improved QueryPath:

1. [Linked the existing PHPUnit tests to GitHub Actions](https://github.com/GravityPDF/querypath/commit/6261fbad21b84b3e49c5ff951e352fd8ed5dbbfd), so the test suite is run automatically on every commit / PR in PHP7.1 up to PHP8.1
2. [Setup PSR-2-inspired linting rules](https://github.com/GravityPDF/querypath/commit/03bd2c3a94094a67198cd7dd5bf6bf0ed9a93dce), linked it to GitHub Actions to automatically test all new commits, and applied the rules to the entire codebase.
3. [Added PHP8.1 support to the library](https://github.com/GravityPDF/querypath/commit/6261fbad21b84b3e49c5ff951e352fd8ed5dbbfd)([here too](https://github.com/GravityPDF/querypath/commit/be4055d25931e49d473d633623d1a5f534176879)).
5. Made preparations to help facilitate collaboration with other users through standardized Issue and Pull Request templates in GitHub, contributors guide, security notice, and changelog file, to name a few. 
6. Improved the README to help people get started with QueryPath quicker, and guide them to places they can go to find out more, or contribute back
7. [Started looking at PHP8.2 support](https://github.com/GravityPDF/querypath/pull/26)

What we won't be doing is rewriting the library for the sake of rewriting it (who has the time). We'll iterate as needed to ensure the primary goals are met. Over time the secondary goals should become a byproduct of this work. 

Any questions, let me know. 

Cheers!

**Reply by technosophos** [reply]
That all sounds great. Would it be useful for me to put an official notice at the top of my QueryPath GH repo point users to this fork as the "maintained version"?

**Reply by jakejackson1** [reply]
@technosophos that would be fantastic. Thank you!

**Reply by technosophos** [reply]
Okay, I have added a big header on the top of the original QueryPath page. let me know if there is anything there you want me to change.

I'll also look into how I have querypath.org set up and see if I can redirect that to another site. I believe I was using Doxygen to generate the API docs. But chances are the setup I had back then is not going to work with a modern version.

**Comment by jakejackson1** [comment]
Looks great @technosophos  👍

Once I [get through the example refactoring](https://github.com/GravityPDF/querypath/pull/36/files), I planned to look more closely at the [phing build](https://github.com/GravityPDF/querypath/blob/main/build.xml#L35) and [makefile](https://github.com/GravityPDF/querypath/blob/main/Makefile) to see what is still usable. 

If I can get Doxygen (or something similar) to create the API docs I'll likely host it with GitHub pages. I can let you know once that is up if you still wanted to add a redirect from querypath.org.

**Reply by technosophos** [reply]
Sounds good. I'll keep an eye out.
